test/Microsoft.AspNetCore.Server.Kestrel.Core.Tests/RequestBodyReaderTests.cs

```diff
@@ -0,0 +1,206 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+using Microsoft.Extensions.Internal;
+using Moq;
+using Xunit;
+using Xunit.Sdk;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Tests
+{
+    public class RequestBodyReaderTests
+    {
+        [Theory]
+        [MemberData(nameof(RequestData))]
+        public async Task CopyToAsyncDoesNotCopyBlocks(FrameRequestHeaders headers, string[] data)
+        {
+            var writeCount = 0;
+            var writeTcs = new TaskCompletionSource<byte[]>();
+
+            var mockDestination = new Mock<Stream>();
+            mockDestination
+                .Setup(m => m.WriteAsync(It.IsAny<byte[]>(), It.IsAny<int>(), It.IsAny<int>(), CancellationToken.None))
+                .Callback((byte[] buffer, int offset, int count, CancellationToken cancellationToken) =>
+                {
+                    writeTcs.SetResult(buffer);
+                    writeCount++;
+                })
+                .Returns(TaskCache.CompletedTask);
+
+            using (var input = new TestInput())
+            {
+                var body = MessageBody.For(HttpVersion.Http11, headers, input.FrameContext);
+                var reader = new RequestBodyReader(input.PipeFactory.Create());
+                var readerTask = reader.StartAsync(body);
+
+                var copyToAsyncTask = reader.CopyToAsync(mockDestination.Object);
+
+                var pipe = input.Pipe;
+                var bytes = Encoding.ASCII.GetBytes(data[0]);
+                var buffer = pipe.Writer.Alloc(2048);
+                Assert.True(buffer.Buffer.TryGetArray(out var block));
+                Buffer.BlockCopy(bytes, 0, block.Array, block.Offset, bytes.Length);
+                buffer.Advance(bytes.Length);
+                await buffer.FlushAsync();
+
+                // Verify the block passed to WriteAsync is the same one incoming data was written into.
+                Assert.Same(block.Array, await writeTcs.Task);
```

Comment:
I think this test is passing only because you're using the same PipeFactory and therefore the same MemoryPool for  input.Pipe and RequestBodyReader._pipe. Since both Pipes are backed by the same pool, the data copied from input.Pipe to the writable buffer of RequestBodyReader._pipe ends up in the same array instance.

After this change, CopyToAsyncDoes~~Not~~CopyBlocks, so maybe this test should just be removed.